src/diagram.rs

```diff
@@ -0,0 +1,771 @@
+    ///     "start": "split",
+    ///     "ops": {
+    ///         "split": {
+    ///             "type": "split",
+    ///             "index": ["op1", "op2"]
+    ///         },
+    ///         "op1": {
+    ///             "type": "node",
+    ///             "builder": "foo",
+    ///             "next": "join"
+    ///         },
+    ///         "op2": {
+    ///             "type": "node",
+    ///             "builder": "bar",
+    ///             "next": "join"
+    ///         },
+    ///         "join": {
+    ///             "type": "join",
+    ///             "inputs": ["op1", "op2"],
+    ///             "next": { "builtin": "terminate" }
+    ///         }
+    ///     }
+    /// }
+    /// # "#)?;
+    /// # Ok::<_, serde_json::Error>(())
+    /// ```
+    Join(JoinOp),
+
+    /// If the request is serializable, transform it by running it through a [CEL](https://cel.dev/) program.
+    /// The context includes a "request" variable which contains the request.
+    ///
+    /// # Examples
+    /// ```
+    /// # bevy_impulse::Diagram::from_json_str(r#"
+    /// {
+    ///     "version": "0.1.0",
+    ///     "start": "transform",
+    ///     "ops": {
+    ///         "transform": {
+    ///             "type": "transform",
+    ///             "cel": "request.name",
+    ///             "next": { "builtin": "terminate" }
+    ///         }
+    ///     }
+    /// }
+    /// # "#)?;
+    /// # Ok::<_, serde_json::Error>(())
+    /// ```
+    ///
+    /// Note that due to how `serde_json` performs serialization, positive integers are always
+    /// serialized as unsigned. In CEL, You can't do an operation between unsigned and signed so
+    /// it is recommended to always perform explicit casts.
+    ///
+    /// # Examples
+    /// ```
+    /// # bevy_impulse::Diagram::from_json_str(r#"
+    /// {
+    ///     "version": "0.1.0",
+    ///     "start": "transform",
+    ///     "ops": {
+    ///         "transform": {
+    ///             "type": "transform",
+    ///             "cel": "int(request.score) * 3",
+    ///             "next": { "builtin": "terminate" }
+    ///         }
+    ///     }
+    /// }
+    /// # "#)?;
+    /// # Ok::<_, serde_json::Error>(())
+    /// ```
+    Transform(TransformOp),
+
+    /// Drop the request, equivalent to a no-op.
+    Dispose,
+}
+
+type DiagramStart = serde_json::Value;
+type DiagramTerminate = serde_json::Value;
+type DiagramScope<Streams = ()> = Scope<DiagramStart, DiagramTerminate, Streams>;
+
+/// Returns the schema for [`String`]
+fn schema_with_string(gen: &mut schemars::gen::SchemaGenerator) -> schemars::schema::Schema {
+    gen.subschema_for::<String>()
+}
+
+/// deserialize semver and validate that it has a supported version
+fn deserialize_semver<'de, D>(de: D) -> Result<semver::Version, D::Error>
+where
+    D: serde::Deserializer<'de>,
+{
+    let s = String::deserialize(de)?;
+    let ver_req = semver::VersionReq::parse(SUPPORTED_DIAGRAM_VERSION).unwrap();
```

Comment:
This seems like it should return an error rather than unwrapping. If we can guarantee that the unwrap won't panic then we need to include a comment providing the rationale.